### PR TITLE
Implement Guard::safepoint()

### PR DESCRIPTION
**Pull request by jeehoonkang**
This is the implementation of `Guard::safepoint()` we discussed in https://github.com/crossbeam-rs/rfcs/issues/18.

I don't expect it to be merged before 0.1.0, but I'm personally in favor of doing so; I need it in implementing a wait-free queue.

### CODE REVIEW

File: src/guard.rs [hunk 1]
```diff
@@ -192,6 +192,38 @@ impl Guard {
             local.flush(self);
         }
     }
+
+    /// Unpins and then pins the current thread.
+    ///
+    /// Since `&mut self` is given, we can statically deduce that there is no [`Shared`] pointer
+    /// that is created before an invocation of `safepoint()` and then survives after the
+    /// invocation.
+    ///
```
Comment (ghost) [inline src/guard.rs]: Should probably also mention that the current thread will be repinned only if this is the only active guard.

**Review by ghost — APPROVED** [review]
LGTM after adding a comment.

**Comment by ghost** [comment]
By the way, do you think your queue would be better served with HP instead of EBR for memory reclamation? The published paper says they used HP - probably for a good reason?

**Comment by jeehoonkang** [comment]
I just rebased and added a comment.

I think HP is more friendly for wait-free constructions.  In almost every wait-free construction, a data structure (usually linked list) contains all the threads, and a thread has a pointer to another thread in the data structure.  "A points to B" means that A will help B in the next turn.  That help pointer is usually long-living: it is updated only if A succeeds an operation, and the pointer points to a valid object until an iteration of the data structure is over.  So If this pointer is managed by EBR, the global epoch will be very slowly advanced.

I'm trying to deal with long-living pointers with crossbeam-epoch by designating an independent collector that only deals with the help pointers.  It's fine to slowly advance its epoch, because there will not be so many help-pointers.  (WARNING: it's not finished, so I still don't know whether it works.)

I'm also interested in mixing HP and EBR as discussed in https://github.com/crossbeam-rs/rfcs/issues/8.  Maybe MSR's Snowflake is an interesting example of that direction: https://www.microsoft.com/en-us/research/publication/project-snowflake-non-blocking-safe-manual-memory-management-net/

**Comment by Amanieu** [comment]
I actually implemented something similar: https://github.com/Amanieu/crossbeam-epoch/commit/7f535bdb57bbb5c8f5b99a3b48b07be1cfc1c533

File: src/internal.rs [hunk 2]
```diff
@@ -283,14 +288,30 @@ impl Local {
...
+    pub fn repin(&self) {
+        let guard_count = self.guard_count.get();
+
+        // Update the local epoch only if there's only one guard.
+        if guard_count == 1 {
+            let epoch = self.epoch.load(Ordering::Relaxed);
+            let global_epoch = self.global().epoch.load(Ordering::Relaxed);
+
+            if epoch != global_epoch {
+                self.update_epoch(epoch, global_epoch, Ordering::Release);
```
Comment (Amanieu) [inline src/internal.rs]: A `SeqCst` fence isn't required here, a `Release` store should be enough to ensure correct ordering.

- We need to ensure any loads/stores from the previous epoch do not leak into the new one.
- However we don't care if any loads/store from the new epoch happen before the epoch counter is updated. At worse , other threads will see us in an older epoch and will delay GC slightly.

Comment (jeehoonkang) [reply src/internal.rs]: @Amanieu you're right.  Thanks for pointing out that!  Also it seems I forgot what I said a month ago: https://github.com/crossbeam-rs/rfcs/issues/18#issuecomment-336768214  😭

**Comment by jeehoonkang** [comment]
I rebased and addressed @Amanieu's comment.  Thanks!

**Comment by Amanieu** [comment]
I think the consensus from #45 is that this function should be renamed to `repin`.

**Comment by jeehoonkang** [comment]
I rebased and changed the comment for `repin()`.

File: src/guard.rs [hunk 3]
```diff
@@ -203,10 +240,6 @@ impl Guard {
     ///
     /// If this method is called from an [`unprotected`] guard, then the passed function is called
     /// directly without unpinning the thread.
-    ///
-    /// # Examples
-    ///
-    /// ```
```
Comment (Amanieu) [inline src/guard.rs]: You accidentally deleted part of a comment here.

**Pull request merged**